compose/config.py

```diff
@@ -189,8 +189,20 @@ def merge_service_dicts(base, override):
             override.get('volumes'),
         )
 
+    list_keys = ['ports', 'expose', 'external_links']
+
+    for key in list_keys:
+        if key in base or key in override:
+            d[key] = base.get(key, []) + override.get(key, [])
+
+    list_or_string_keys = ['dns', 'dns_search']
+
+    for key in list_or_string_keys:
+        if key in base or key in override:
+            d[key] = to_list(base.get(key)) + to_list(override.get(key))
+
     for k in ALLOWED_KEYS:
-        if k not in ['environment', 'volumes']:
+        if k not in ['environment', 'volumes'] + list_keys + list_or_string_keys:
```

Comment:
Minor: this might be easier to read as:

```python
already_merged_keys = set(['environment', 'volumes'] + list_keys + list_or_string_keys)
for k in set(ALLOWED_KEYS) - already_merged_keys:
    ...

```